chore: add v4 migration script for adapters

ndom91 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: https://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

## Reasoning 💡

<!-- What changes are being made? What feature/bug is being fixed here? -->

Initial attempt at some DB migration scripts in the `v4 upgrade` docs. For migrating adapters created under v3 to the new v4 schema.

## Checklist 🧢

<!-- Feel free cross items ( like this `~[] item~` ) if they're irrelevant to your changes.

To check an item, place an `x` in the box like so: `- [x] Documentation`. -->

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

<!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

## Affected issues 🎟

<!--
Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

If you write `"Fixes"` or `"Closes"` before the issue link like so:

```
Fixes #359
```

the connected issue will be automatically closed once the PR is merged and hence help with maintenance of the library 😊

-->

ndom91 committed:
chore: add v4 migration script for adapters

ndom91 committed:
chore: cleanup

ndom91 committed:
Merge branch 'main' into ndom91/docs/adapter-migration-script

balazsorban44 commented on packages/docs/docs/getting-started/upgrade-to-v4.md:
```diff
@@ -439,6 +439,106 @@ See the [`session` option docs](https://next-auth.js.org/configuration/options#s
+ADD COLUMN "token_type" varchar(191) NULL
+ADD COLUMN "scope" varchar(191) NOT NULL
+ADD COLUMN "id_token" varchar(191) NOT NULL
+ADD COLUMN "oauth_token_secret" varchar(191) NOT NULL
+ADD COLUMN "oauth_token" varchar(191) NOT NULL
```
So these are only needed if they are using Twitter OAuth 1.0. We should comment and link to the docs. Otherwise these can be dropped.

balazsorban44 commented on packages/docs/docs/getting-started/upgrade-to-v4.md:
```diff
@@ -439,6 +439,106 @@ See the [`session` option docs](https://next-auth.js.org/configuration/options#s
+RENAME COLUMN "provider_account_id" "providerAccountId"
+DROP COLUMN "provider_type"
+DROP COLUMN "compound_id"
+DROP COLUMN "created_at"
+DROP COLUMN "updated_at"
```
Let's comment that these have never been necessary for NextAuth.js to work, hence they are dropped, but can be kept if they want to.

balazsorban44 commented on packages/docs/docs/getting-started/upgrade-to-v4.md:
```diff
@@ -439,6 +439,106 @@ See the [`session` option docs](https://next-auth.js.org/configuration/options#s
+
+/* USER */
+ALTER TABLE users
+DROP COLUMN "created_at"
+DROP COLUMN "updated_at"
```
Same as https://github.com/nextauthjs/next-auth/pull/3874/files#r800927573

balazsorban44 commented on packages/docs/docs/getting-started/upgrade-to-v4.md:
```diff
@@ -439,6 +439,106 @@ See the [`session` option docs](https://next-auth.js.org/configuration/options#s
+CHANGE "user_id" "userId" varchar(191)
+ADD CONSTRAINT fk_user_id FOREIGN KEY (user_id) REFERENCES users(id)
+DROP COLUMN "access_token"
+DROP COLUMN "created_at"
+DROP COLUMN "updated_at"
```
https://github.com/nextauthjs/next-auth/pull/3874/files#r800927573

balazsorban44 commented on packages/docs/docs/getting-started/upgrade-to-v4.md:
```diff
@@ -439,6 +439,106 @@ See the [`session` option docs](https://next-auth.js.org/configuration/options#s
+/* VERIFICATION REQUESTS */
+ALTER TABLE verification_requests
+DROP COLUMN id
+DROP COLUMN "created_at"
+DROP COLUMN "updated_at"
```
https://github.com/nextauthjs/next-auth/pull/3874/files#r800927573

balazsorban44 reviewed (changes requested):
Only commented on the MySQL one, but should be the same for Postgres

TheoBr commented:
Looks good overall, really appreciate the separate examples for MySQL and Postgres 

Two concerns:
- “Rename” isn’t an option for a lot of sharded dbs [like Planetscale](https://github.com/planetscale/beta/discussions/122)
- by making multiple new fields non nullable, you're likely breaking any existing entries in those tables

ndom91 committed:
chore: PR comments

ndom91 commented:
Thanks for the feedback everyone! I've addressed the comments so far. 

Regarding `RENAME` not working with sharded DBs like planetscale, I think that basically rules out using an automated migration like this, right. In cases like those, if the user wants to upgrade from v3 -> v4, they'll have to manually adjust the DB if possible or potentially just create a new one and move the data over. I know that sucks, but I don't see another option if `RENAME` isn't available. 

What do yall think?

TheoBr commented:
```
  type               String   @map("providerType")
  refresh_token      String?  @map("refreshToken") @db.Text
  provider           String   @map("providerId")
```
^ is the fix we're using at Ping for now

ndom91 commented:
Yeah just remap the names in Prisma. Makes sense. For those using prisma this seems like a nice workaround, will keep it in mind!

TheoBr commented:
My other concern was w/ the non-nullable bindings to `expires_at` and `session_state`. Neither seem to be used, this screenshot is from a minimal production app on nextauthjs v4

![image](https://user-images.githubusercontent.com/6751787/152879594-790d2bc0-5f45-410d-b466-331601d9f047.png)

ndom91 commented:
@TheoBr yeah I changed those to no longer be non-nullable, sticking to the previous implementation. New v4 setups can keep the non-nullable variants going forward, but for migrations you're right - that would probably break some current implementations :+1: 

Or do you mean to allow NULL on `created_at` + `updated_at` ?

ndom91 committed:
chore(docs): add mongodb migration example + fix nextauthjs/adapters links

ndom91 committed:
Merge branch 'main' into ndom91/docs/adapter-migration-script

balazsorban44 reviewed (approved):
LGTM :heavy_check_mark:
